single_kernel_mongo/events/tls.py

```diff
@@ -9,205 +9,225 @@
+        self.framework.observe(self.charm.on.secret_changed, self._on_secret_changed)
+
+    @property
+    def tls_mapping(self) -> dict[bool, TLSCertificatesRequiresV4]:
+        """Mapping of boolean to a TLS requirer instance.
+
+        The boolean value is True if the requirer is for internal certificates,
+        and False for the client certificates.
+        """
+        return {True: self.peer_certificate, False: self.client_certificate}
+
+    def _on_tls_relation_created(self, event: RelationCreatedEvent) -> None:
+        """Handler for relation created."""
+        if event.relation.name != ExternalRequirerRelations.PEER_TLS.value:
             return
-        if self.manager.state.upgrade_in_progress:
-            logger.warning(
-                "Enabling TLS is not supported during an upgrade. The charm may be in a broken, unrecoverable state."
+        if self.manager.state.is_role(MongoDBRoles.MONGOS):
+            self.manager.state.statuses.delete(
+                MongosStatuses.MISSING_PEER_TLS_REL.value,
+                scope="unit",
+                component=self.dependent.name,
             )
-            event.defer()
-            return
 
-        # When we can integrate, clean the mongos requires tls status.
-        if self.manager.state.is_role(MongoDBRoles.MONGOS):
+        if self.manager.state.is_role(MongoDBRoles.SHARD):
             self.manager.state.statuses.delete(
-                MongosStatuses.MISSING_TLS_REL.value, scope="unit", component=self.dependent.name
+                ShardStatuses.MISSING_PEER_TLS_REL.value,
+                scope="unit",
+                component=self.dependent.name,
             )
 
-        for internal in (True, False):
-            csr = self.manager.generate_certificate_request(None, internal=internal)
-            self.certs_client.request_certificate_creation(certificate_signing_request=csr)
-            self.manager.set_waiting_for_cert_to_update(internal=internal, waiting=True)
+    def refresh_certificates(self) -> None:
+        """Trigger refresh TLS certificates event."""
+        logger.info(f"Requesting refresh certificates for unit: {self.charm.unit.name}.")
+        self.refresh_tls_certificates_event.emit()
 
     def _on_tls_relation_broken(self, event: RelationBrokenEvent) -> None:
-        """Handler for relation joined."""
-        if not self.manager.state.db_initialised:
-            logger.info(f"Deferring {str(type(event))}. db is not initialised.")
-            event.defer()
-            return
+        """Handle the relation broken event."""
+        state = self.manager.get_tls_management_state()
+        match state:
+            case (
+                TlsManagementState.DB_NOT_INTIALIZED
+                | TlsManagementState.MONGOS_DB_NOT_INITIALIZED
+                | TlsManagementState.UPGRADE_IN_PROGRESS
+            ):
+                defer_event_with_info_log(logger, event, str(type(event)), state.value)
```

Comment:
why should we defer if the TLS relation is broken and the DB is not initialized?